hedera-node/hedera-token-service-impl/src/main/java/com/hedera/node/app/service/token/impl/handlers/TokenAccountWipeHandler.java

```diff
@@ -54,8 +84,123 @@ public void preHandle(@NonNull final PreHandleContext context) throws PreCheckEx
         }
     }
 
+    @Override
+    public void pureChecks(@NonNull final TransactionBody txn) throws PreCheckException {
+        final var op = txn.tokenWipeOrThrow();
+
+        // All the pure checks for burning a token must also be checked for wiping a token
+        burnPureChecks(op.amount(), op.serialNumbers(), op.hasToken(), INVALID_WIPING_AMOUNT);
+
+        validateTruePreCheck(op.hasAccount(), INVALID_ACCOUNT_ID);
+    }
+
     @Override
     public void handle(@NonNull final HandleContext context) throws HandleException {
-        throw new UnsupportedOperationException("Not implemented");
+        requireNonNull(context);
+        // Set up the stores and helper objects needed
+        final var accountStore = context.writableStore(WritableAccountStore.class);
+        final var tokenStore = context.writableStore(WritableTokenStore.class);
+        final var tokenRelStore = context.writableStore(WritableTokenRelationStore.class);
+        final var nftStore = context.writableStore(WritableNftStore.class);
+        final var expiryValidator = context.expiryValidator();
+
+        // Assign relevant variables
+        final var txn = context.body();
+        final var op = txn.tokenWipeOrThrow();
+        final var accountId = op.account();
+        final var tokenId = op.token();
+        final var fungibleBurnCount = op.amount();
+        // Wrapping the serial nums this way de-duplicates the serial nums:
+        final var nftSerialNums = new ArrayList<>(new LinkedHashSet<>(op.serialNumbers()));
+
+        // Validate the semantics of the transaction
+        final var validated = validateSemantics(
+                accountId,
+                tokenId,
+                fungibleBurnCount,
+                nftSerialNums,
+                accountStore,
+                tokenStore,
+                tokenRelStore,
+                expiryValidator);
+        final var acct = validated.account();
+        final var token = validated.token();
+
+        final long newTotalSupply;
+        final long newAccountBalance;
+        final Account.Builder updatedAcctBuilder = validated.account().copyBuilder();
+        if (token.tokenType() == TokenType.FUNGIBLE_COMMON) {
+            // Check that the new total supply will not be negative
+            newTotalSupply = token.totalSupply() - fungibleBurnCount;
+            validateTrue(newTotalSupply >= 0, INVALID_WIPING_AMOUNT);
+
+            // Check that the new token balance will not be negative
+            newAccountBalance = validated.accountTokenRel().balance() - fungibleBurnCount;
+            validateTrue(newAccountBalance >= 0, INVALID_WIPING_AMOUNT);
+        } else {
+            // Check that the new total supply will not be negative
+            newTotalSupply = token.totalSupply() - nftSerialNums.size();
+            validateTrue(newTotalSupply >= 0, INVALID_WIPING_AMOUNT);
```

Comment:
Do we have these checks in `mono-service` for NFT wipe ?

Reply:
This one specifically doesn't seem to be in `Token.wipe(...)` (NFT version), but seemed like it should be there